### ISSUE TITLE
Funny behavior when initializing variable with built-in generic type and contradicting split initialization.

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#19226]
If I try and compile and run the following:

```chpl
use IO;
var r = read(int);

var x : numeric;
if(r > 2) {
  x = 2;
} else {
  x = 3.0i;
}
writeln(x.type : string);
```

I'll get this error:

`foo.chpl:6: error: cannot initialize 'x' of type 'imag(64)' from '2'`.

To which I'm tempted to say: "but I told you `x` is a `numeric` not a `imag(64)`!" 

So I assume when type inference sees something given a built in generic type like "numeric" it then looks at subsequent assignments to see if it can "lock in" some concrete type.  This seems fair enough (Chapel is a strongly and statically typed language), but there are a couple things I find odd about this:

(1) It errors out in the `x=2` statement in the `if` which is lexically before the assignment in the `else`.  I guess type inference or whatever analysis is driving this steps into the 'else' block first. This could get especially confusing if that "first" assignment to `x` weren't so evident but rather buried in a mountain of code somewhere (maybe assigned to the result of a function call or passed to a function as a ref argument).

(2) If I do the following and use split initialization with an (initially) untyped variable I get a different error message. Actually I get two error messages, but I kind of like it because it tells me what lines the contradicting initializations are on:

```chpl
use IO;
var r = read(int);

var x;
if(r > 2) {
  x = 2;
} else {
  x = 3.0i;
}
writeln(x.type : string);
```

```
foo.chpl:8: error: Split initialization uses multiple types; another initialization has type int(64) but this initialization has type imag(64)
foo.chpl:6: error: cannot initialize 'x' of type 'imag(64)' from '2'
```

Anyway, why bring this up? There are a few possible improvements I could see:

* When trying to come up with a type for `x` could we switch things to analyze the 'if' body first? Maybe there's a good reason why we do things in that order?
* Could we change the error message to print out the location of the contradicting assignments?
* Do we explain this behavior in the doc somewhere (that if you declare a variable with a built-in generic type, once you assign it to some value with a concrete type that type is "locked" in).  I came up with this example because I was reading the doc on types, got to the third sentence in the section about Primitive types (https://chapel-lang.org/docs/language/spec/types.html#primitive-types), which linked to the section on "Built-in generic types" (https://chapel-lang.org/docs/language/spec/generics.html#built-in-generic-types), started reading that and my first thought was "hey, are these types generic at runtime?". So maybe we'd want to add something there.

**Comment by bradcray** [comment]
I think you're right that this behavior is strange/surprising, and would expect the two examples you give to behave more like one another (and ideally more like the second).  That said, I think my ideal error message would be something like

foo.chpl:4: error: Can't infer the type of 'x' because we can't unify
foo.chpl:6: note: type 'int(64)' with
foo.chpl:8: note: type 'imag(64)'.
```

(focusing more on not picking sides for 'x' and pointing to the cases that won't unify rather than the specific wording or formatting given here)

**Comment by mppf** [comment]
I like the concept of @bradcray's error message proposal above but I'm not sure if "unify" is a term that will immediately make sense to users encountering this situation.

**Comment by stonea** [comment]
> I poked around in gdb and I think it does visit the if first, but then also visits the else (and following branches) so the "last resolve wins". Including a three-arm conditional here to demonstrate. The output prints the type name that the variable has at each change.

Huh, given your description I was hoping this meant we gathered all the types and then unified them to the broadest type at the end. But you're right, it looks like it just takes the last type and if that doesn't work with everything else it errors out.

So this will work fine and type `x` as an `int(16)`:

```chpl
use IO;
writeln("Enter a value: ");
var r = read(int);

var x : numeric;
if(r == 0) {
  x = 1 : int(8);
} else {
  x = 1024 : int(16);
}
writeln("x is a ", x.type : string, " assigned to ", x);
```

And the following errors out with "foo.chpl:7: error: cannot initialize 'x' of type 'int(8)' from '1024: int(16)'"

```chpl
use IO;
writeln("Enter a value: ");
var r = read(int);

var x : numeric;
if(r == 0) {
  x = 1024 : int(16);
} else {
  x = 1 : int(8);
}
writeln("x is a ", x.type : string, " assigned to ", x);
```

Ideally I'd like it if we would just infer `x` to be an `int(16)` in both cases. 

With my first example (with `imag` and `int`) I think Brad's error message about being unable to unify the two types would be better than what we have, but with the second example I think I'd be especially confused. At least I don't think of type unification as being a directional thing (though maybe my understanding of it is wrong).

**Comment by bradcray** [comment]
> At least I don't think of type unification as being a directional thing

I agree that it would be preferable to not have unification be based on the order in which the compiler saw expressions.  I.e., I should be able to write my conditionals as `if cond then ... else ...` or `if !cond then ... else ...` without worrying about it affecting the code's correctness or type inference decisions.

**Comment by mppf** [comment]
I think the main thing I am wondering about this is whether or not we want to fix the production resolver or live with it for a while and fix it in the dyno resolver. I do not think we are very far from working on split-init type resolution there but it will take some time for us to start using the dyno resolver by default (in particular I would not expect that in 1.26).

**Comment by bradcray** [comment]
I would personally be fine waiting to fix this in the new compiler.  At least for awhile.

**Comment by stonea** [comment]
This issue isn't a blocker for me or anything (just something I noticed while playing around with Types for the module review) and there's a straight-forward workaround (just manually type things the way like them to be inferred).

So I think makes sense to wait and not duplicate effort.

**Comment by mppf** [comment]
So, I am thinking that if type inference is used with split initialization, then we should insist that the types match exactly, rather than trying to unify/find the common type. The reason is that I would like the resolver to be able to proceed in order.

Consider, for example, this program:

``` chapel
config const cond = true;

proc main() {
  var x;

  if cond {
    x = 5;
    compilerWarning('in if branch, x.type is ' + x.type:string);
  } else {
    x = 12.0;
    compilerWarning('in else branch, x.type is ' + x.type:string);
  }

  compilerWarning('afterwards, x.type is ' + x.type:string);
}
```

If the resolver is proceeding in order, it will process the `if` branch first. It will see `x = 5` and then infer that `x` must have type `int`. But then, when processing the `else` branch, it has in mind that `x` is `int` but now sees a `real`. It could "unify" and change the type of `x` to `int`, but it has already resolved code using the type of `x`.

You can see this happening with the production resolver:
```
bb.chpl:3: warning: in if branch, x.type is int(64)
bb.chpl:3: In function 'main':
bb.chpl:10: error: Split initialization uses multiple types; another initialization has type int(64) but this initialization has type real(64)
bb.chpl:3: warning: in else branch, x.type is real(64)
bb.chpl:3: warning: afterwards, x.type is real(64)
```

I think it's OK to temporarily be wrong about a type like this if there is going to be a compilation error. But if we were to make this case not be a compilation error, then `x` would have different types within each conditional -- or we would have to give up on resolving these in order (and that seems like something that would introduce a lot of other problems).

So, my conclusion is, split-init for conditionals should require that the types match exactly. We can certainly improve the error message, however.

From the OP:

> Do we explain this behavior in the doc somewhere 

In https://chapel-lang.org/docs/language/spec/variables.html#split-initialization it says this:

> In the case that the variable is declared without a type-part and where multiple applicable assignment statements are identified, all of the assignment statements need to contain an initialization expression of the same type.

This could be improved to say "declared without a type-part or with a generic type". Also I think it would help the spec if we make Split Initialization a top-level section in Variables since as it is, it's pretty buried in the docs.

**Comment by bradcray** [comment]
That would be OK with me.  It has the virtue of being relaxable later if we came up with compelling reasons to do so, whereas choosing other options would result in backwards breaking changes if we decided this way was the right way forward later on.
